### PR TITLE
[sanitizer] Add MemCpyAccessible

### PR DESCRIPTION
**Pull request by vitalybuka**
A layer over `TryMemCpy` to copy only availible
pages.

### CODE REVIEW

File: compiler-rt/lib/sanitizer_common/sanitizer_common.h [hunk 1]
```diff
@@ -273,6 +273,8 @@ bool IsAccessibleMemoryRange(uptr beg, uptr size);
 // Returns true if we can read a memory range starting at `src`, and copies
 // content into `dest`.
 bool TryMemCpy(void *dest, const void *src, uptr n);
+// Copies accesible memory, and zero fill the rest.
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_common.h:276]: nit: typo

File: compiler-rt/lib/sanitizer_common/sanitizer_common_libcdep.cpp [hunk 2]
```diff
@@ -219,6 +219,32 @@ static void StopStackDepotBackgroundThread() {
...
+    return;
+
+  const uptr page_size = GetPageSize();
+  uptr b = reinterpret_cast<uptr>(src);
+  uptr b_up = RoundUpTo(b, page_size);
+
+  uptr e = reinterpret_cast<uptr>(src) + n;
+  uptr e_down = RoundDownTo(e, page_size);
+
+  auto copy_or_zero = [dest, src](uptr b, uptr e) {
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_common_libcdep.cpp:233]: Don't reuse `b` and `e` in and outside of the lambda

File: compiler-rt/lib/sanitizer_common/sanitizer_common_libcdep.cpp [hunk 3]
```diff
@@ -219,6 +219,32 @@ static void StopStackDepotBackgroundThread() {
 static void StopStackDepotBackgroundThread() {}
 #endif
 
+void MemCpyAccessible(void *dest, const void *src, uptr n) {
+  if (TryMemCpy(dest, src, n))
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_common_libcdep.cpp:223]: I am not fully sure about this optimization. In the case where there are holes this makes it more expensive. Does the caller generally know that?

Comment (vitalybuka) [reply compiler-rt/lib/sanitizer_common/sanitizer_common_libcdep.cpp:223]: Not sure as well. On some examples fast pass 4x faster than page by page.

File: compiler-rt/lib/sanitizer_common/sanitizer_common_libcdep.cpp [hunk 4]
```diff
@@ -230,18 +230,18 @@ void MemCpyAccessible(void *dest, const void *src, uptr n) {
...
 
-  auto copy_or_zero = [dest, src](uptr b, uptr e) {
-    uptr d = reinterpret_cast<uptr>(dest) + (b - reinterpret_cast<uptr>(src));
-    if (!TryMemCpy(reinterpret_cast<void *>(d), reinterpret_cast<void *>(b),
-                   e - b))
-      internal_memset(reinterpret_cast<void *>(d), 0, e - b);
+  const uptr off = reinterpret_cast<uptr>(dest) - b;
+
+  auto copy_or_zero = [off](uptr beg, uptr end) {
+    void *d = reinterpret_cast<void *>(beg + off);
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_common_libcdep.cpp:236]: i preferred the previous calculation, it's weird to have the offset between source and destination.

File: compiler-rt/lib/sanitizer_common/tests/sanitizer_posix_test.cpp [hunk 5]
```diff
@@ -127,6 +127,43 @@ TEST(SanitizerCommon, TryMemCpyNull) {
...
+  std::vector<char> dst;
+  std::vector<char> exp = {src.begin(), src.end()};
+
+  // Protect some pages.
+  for (int i = 7; i < page_num; i *= 2) {
+    mprotect(src.data() + i * page_size, page_size, PROT_NONE);
+    std::fill(exp.data() + i * page_size, exp.data() + (i + 1) * page_size, 0);
+  }
+
+  dst.assign(src.size(), 0);
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/tests/sanitizer_posix_test.cpp:144]: nit '\0'

**Review by fmayer — APPROVED** [review]

**Pull request merged**